private_sharing/api_views.py

```diff
@@ -458,3 +470,92 @@ def post(self, request):
+
+    Accepts project name and description as (required) inputs
+
+    A third input that should be provided is the first part of the redirect url;
+    this will get the new project's slug appended to it to form the new project's
+    oauth2 redirect url, eg <mysuperspiffydomain>/diyprojects/<new-project-slug>/complete/
+
+    The other required fields are auto-populated:
+    is_study:  set to False
+    leader:  set to member.name from oauth2 token
+    coordinator:  get from oauth2 token
+    is_academic_or_nonprofit: False
+    add_data:  false
+    explore_share:  false
+    short_description:  first 139 chars of long_description plus an elipse
+    active:  True
+    coordinator:  from oauth2 token
+    """
+
+    authentication_classes = (CustomOAuth2Authentication,)
+    permission_classes = (HasValidProjectToken,)
+
+    def get_short_description(self, long_description):
+        """
+        Return first 139 chars of long_description plus an elipse.
+        """
+        return "{0}…".format(long_description[0:139])
+
+    def post(self, request):
+        """
+        Take incoming json and create a project from it
+        """
+        project_creation_project = OAuth2DataRequestProject.objects.get(
+            pk=self.request.auth.pk
+        )
+
+        # If the first part of the redirect_url is provided, grab that, otherwise set
+        # to the project-creation-project's enrollment_url as a usable default
+        redirect_url_part = request.data.get("redirect-url-part", None)
+        if not redirect_url_part:
+            redirect_url_part = project_creation_project.enrollment_url
+
+        member = get_oauth2_member(request).member
+        serializer = ProjectCreationSerializer(data=request.data)
+        if serializer.is_valid():
+            project = serializer.save(
+                is_study=False,
+                is_academic_or_nonprofit=False,
+                add_data=False,
+                explore_share=False,
+                active=True,
+                short_description=self.get_short_description(
+                    serializer.validated_data["long_description"]
+                ),
+                coordinator=member,
+                leader=member.name,
+                request_username_access=False,
+                diy_project=True,
+            )
+
+            # Coordinator join project
+            project_member = project.project_members.create(member=member)
+            project_member.joined = True
+            project_member.authorized = True
+            project_member.save()
+
+            # Generate redirect URL and save to project
+            project.redirect_url = "{0}/{1}/complete/".format(
+                redirect_url_part, project.slug
+            )
+            project.save()
+
+            # Serialize project data for response
+            # Copy data dict so that we can easily append fields
+            serialized_project = ProjectDataSerializer(project).data
```

Comment:
This is redundant with the project data endpoint, which is OK -- but it's also insufficient.

That serializer does not include client_id or client_secret -- and it can't include the latter, for security reasons. The API needs to also return both of those things -- otherwise the other app can't manage the project on the coordinator's behalf, going forward.

Reply:
Oh, I'd had those in my mockup, but somehow missed including them here.  Oops!